0005-R-pubsub-delayed-scheduling.md

```diff
@@ -0,0 +1,156 @@
+
+### Related issues 
+
+https://github.com/dapr/dapr/issues/2675
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+
+This proposal outlines a new metadata key for the [Dapr publish endpoint](https://docs.dapr.io/reference/api/pubsub_api/#publish-a-message-to-a-given-topic) to apply for both HTTP and gRPC APIs that allows developers to set the target date and time for a message to be published to a given topic. 
+
+* What is deliberately *not* in scope?
+
+Using component specific implementations is deliberately not in scope, as it offers no tangible benefits and carries heavy maintenance burden for contributors, approvers and maintainers.
+
+* What alternatives have been considered, and why do they not solve the problem?
+
+The only alternative for users is to create separate topics for delayed messages together with boilerplate code to orchestrate sending/receiving messages from said topics within their code, and then chaining the message flow to the correct topics after the conditions for sending delayed messages have been met. In addition, user code will need to become aware of Dapr specific mechanisms and mirror it in user code.
+
+* What advantages / disadvantages does this proposal have?
+
+This proposal allows delayed message scheduling to be enabled for:
+
+1. Every pub/sub component supported by Dapr without requiring component specific changes
+2. Messages sent using Dapr CloudEvents
+3. Messages sent using user provided CloudEvents
+4. Messages sent using raw payloads
+5. Message sent/received by HTTP and gRPC subscribers
+
+## Implementation Details
+
+### Design
+
+The Dapr runtime will use an intermediary topic to store messages that are scheduled for late delivery. Users will specify their intent for a message to be delayed using the existing metadata mechanism (specified as a query param on HTTP1.1 or on the request object for gRPC), the same way we use metadata for TTLs in pub/sub and for many other features. The proposed metadata key to be used is `scheduledSendTime` with the value being a date time string adhering to the ISO-8601 format, the same format supported by Dapr actors reminders and timers.
+
+HTTP example:
+
+```bash
+curl -X "POST" http://localhost:3500/v1.0/publish/pubsub/singularity?metadata.scheduledSendTime=2045-11-05T08:15:30-05:00 -H "Content-Type: application/json" -d '{"execute-order": "66"}'
+```
+
+Upon receiving a delayed message, the Dapr runtime will examine the due date for the message and publish it to the target topic if time is due. If the time isn't due, the Dapr runtime will hold the message until the time is right to send it without ACKing back to the broker. If the Dapr instance crashes, the message will remain in the broker and be consumed by a different instance.
```

Comment:
This will not work in every broker. Not acking a message can cause no other messages to be consumed in some configurations. In Kafka, there is in-order guarantee per partition, so not acking a message will clog the partition for other messages - even for those are already due.

Instead, this should implement an outbox pattern instead. In the outbox pattern, messages are only published after a business transaction takes place - in this case, the business transaction is done by the clock. Then, the component must scan states in a state store to identify which messages are due and publish those, deleting the corresponding record.

The state store keys would basically be in segments of 1h (or any other granularity) and contain the ids for every message that must be delivered in that hour. Then, daprd would look at the current hour and previous (configurable by the user to be -n hours) and publish every message that is due in those time windows.

This will make this solution work in a predictable way without depending on broker-specific handling of "not acking messages" by just relying on the state store pattern underneath.

Reply:
@yaron2 has this comment been integrated yet? If so can you resolve the conversation :)

Reply:
It hasn't been resolved yet